solidity/contracts/bridge/Frauds.sol

```diff
@@ -0,0 +1,577 @@
+// SPDX-License-Identifier: MIT
+
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity ^0.8.9;
+
+import {BytesLib} from "@keep-network/bitcoin-spv-sol/contracts/BytesLib.sol";
+import {BTCUtils} from "@keep-network/bitcoin-spv-sol/contracts/BTCUtils.sol";
+import {CheckBitcoinSigs} from "@keep-network/bitcoin-spv-sol/contracts/CheckBitcoinSigs.sol";
+import "./EcdsaLib.sol";
+import "./Bridge.sol";
+
+library Frauds {
+    using BytesLib for bytes;
+    using BTCUtils for bytes;
+    using BTCUtils for uint32;
+    using EcdsaLib for bytes;
+
+    struct Data {
+        ///  The amount of stake slashed from each member of a wallet for a fraud.
+        uint256 slashingAmount;
+        /// The percentage of the notifier reward from the staking contract
+        /// the notifier of a fraud receives. The value is in the range [0, 100].
+        uint256 notifierRewardMultiplier;
+        /// The amount of time the wallet has to defeat a fraud challenge.
+        uint256 challengeDefeatTimeout;
+        /// The amount of ETH the party challenging the wallet for fraud needs
+        /// to deposit.
+        uint256 challengeDepositAmount;
+        /// Collection of all submitted fraud challenges indexed by challenge
+        /// key built as keccak256(walletPublicKey|sighash|v|r|s).
```

Comment:
We can do that (use only `walletPublicKey|sighash` as the challenge key).
But it also means we must remove `signature` from `defeatFraudChallenge` and `notifyFraudChallengeTimeout`, because if the `signature` is no longer part of the challenge key, we are not sure if the values of `signature` passed into these two functions are correct.

We can remove `signature` from the two mentioned functions as it is only used there (apart from building the challenge key) during event emitting.

@lukasz-zimnoch Would you remove the `signature` from `defeatFraudChallenge` and `notifyFraudChallengeTimeout` and not emit the `signature` in the events?

Reply:
I think it makes sense but maybe I am missing some scenario. Only `submitFraudChallenge` is validating the signature. Once we know the wallet unlocked the given UTXO and we saved the preimage, we need the wallet to prove this was an honest unlock. We do not need to validate signature for that.

Reply:
Yes, we only need signature in `submitFraudChallenge`. I removed the signature from building the challenge key in e4fa66ce96d5478fe0eae08a5bd709af7fc2db55.